### PR TITLE
Input refactor for more consistent handling across devices

### PR DESCRIPTION
**Pull request by bitcraft**
Mostly addresses #459 #432:
* input cleanup, unification
* hold B/Cancel button to run

but also minor edits along the way:
* hide_mouse option
* walk/run speed in config file
* gamepad deadzone
* seamless keyboard/controller/gamepad input
* re-configurable inputs at runtime (no ui yet)
* easy checking of buttons held down (for hold-run-run)
* simplified input handling
* pygame is no longer a "hard" dependancy for inputs

Minor changes/cleanup
* linting
* removing some idk, verbose/unneeded code comments
* docstring types

### CODE REVIEW

**Comment by bitcraft** [comment]
@ShadowApex @kerizane  Please take a look when you have a moment.

**Comment by kerizane** [comment]
I'll try to take a look tomorrow

**Comment by kerizane** [comment]
Game crashes whilst trying to name the character on the New Game screen.

**Comment by kerizane** [comment]
Should I be able to do anything with the mouse?

File: tuxemon/core/platform/platform_pygame/events.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+from tuxemon.core.platform.const import buttons, intentions, events
+from tuxemon.core.platform.events import InputHandler, PlayerInput, EventQueueHandler
+
+
+class PygameEventQueueHandler(EventQueueHandler):
+    """ Handle all events from the pygame event queue
+    """
+
+    def __init__(self):
+        self._inputs = defaultdict(list)  # type Dict[int, InputHandler]
```
Comment (kerizane) [inline tuxemon/core/platform/platform_pygame/events.py]: type should be `Dict[int, List[InputHandler]]` ?

File: tuxemon/core/platform/platform_pygame/events.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                if pressed:
+                    self.press(button)
+                else:
+                    self.release(button)
+                return
+
+            # just get unicode value
+            try:
+                if pressed:
+                    print(pg_event.unicode)
```
Comment (kerizane) [inline tuxemon/core/platform/platform_pygame/events.py]: Do we need this print?

File: tuxemon/core/platform/platform_pygame/events.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from collections import defaultdict
+
+import pygame as pg
+
+from tuxemon.core import prepare
+from tuxemon.core.platform.const import buttons, intentions, events
```
Comment (kerizane) [inline tuxemon/core/platform/platform_pygame/events.py]: intentions is unused

**Comment by kerizane** [comment]
Other than that it looks fine.

File: tuxemon/core/platform/platform_pygame/events.py (hunk 2 above)
Comment (bitcraft) [reply tuxemon/core/platform/platform_pygame/events.py]: nope.  ill clean that up

File: tuxemon/core/platform/platform_pygame/events.py (hunk 1 above)
Comment (bitcraft) [reply tuxemon/core/platform/platform_pygame/events.py]: yeah, good catch

File: tuxemon/core/platform/platform_pygame/events.py (hunk 3 above)
Comment (bitcraft) [reply tuxemon/core/platform/platform_pygame/events.py]: ok, ill lint it

**Comment by bitcraft** [comment]
> Should I be able to do anything with the mouse?

Maybe.  I haven't tested mouse use in a long time, but it should be usable with menus.  looks like you've found a regression/bug so ill fix that.

**Comment by bitcraft** [comment]
@kerizane @ShadowApex fixed the mouse regression and addressed the comments

**Comment by kerizane** [comment]
Can't name the character with the keyboard, doesn't crash though.

Trackpad still doesn't work for me. Hard to tell what I'm clicking since the cursor is hidden.

**Comment by bitcraft** [comment]
@kerizane what python and pygame version are you using?  keyboard layout/locale?  direct keyboard might only work for en_us.  mouse needs to be enabled in the options to work.

**Comment by kerizane** [comment]
python3.7, pygame2, en_us.

Will try mouse again later.

**Comment by kerizane** [comment]
Mouse looks fine. Could name character, and select menu items.
Wasn't able to test walking.

**Comment by kerizane** [comment]
OK, weird bug.

Using my bluetooth keyboard the character is stuck moving upwards. Can't get him to stop or move anywhere else.

This isn't the case using a wired keyboard, or a laptop keyboard.
This isn't the case with the old code.

**Comment by bitcraft** [comment]
I used to use that keyboard....  the mouse is great.  If the character is stuck moving, then the KEYUP event got lost somehow.  Are you able to duplicate the problem reliably?

**Comment by kerizane** [comment]
Happens immediately without pressing anything.

**Comment by kerizane** [comment]
If I plug the keyboard in after loading then I can walk around normally.

**Comment by bitcraft** [comment]
wow that is a strange issue.  i fixed the direct keyboard input regression, thanks for pointing that out.

**Comment by bitcraft** [comment]
@kerizane can you go to ```tuxemon/core/platform/platform_pygame/events.py``` and print/log the pygame events and see what the keyboard is sending?

**Comment by kerizane** [comment]
Got it! It thinks it's a gamepad as well as a keyboard.

    def check_axis(self, pg_event):
        if pg_event.type == pg.JOYAXISMOTION:
            value = pg_event.value

            if pg_event.axis == 0:
                print("FIRST THIS")
                if abs(value) >= self.deadzone:
                    if value < 0:
                        self.press(buttons.LEFT, value * -1)
                    else:
                        self.press(buttons.RIGHT, value)
                else:
                    self.release(buttons.LEFT)
                    self.release(buttons.RIGHT)

            elif pg_event.axis == 1:
                if abs(value) >= self.deadzone:
                    if value < 0:
                        print("THEN THIS")
                        self.press(buttons.UP, value * -1)
                    else:
                        print("NOT THIS")
                        self.press(buttons.DOWN, value)
                else:
                    print("NOT THIS")
                    self.release(buttons.UP)
                    self.release(buttons.DOWN)

**Comment by bitcraft** [comment]
looks like you are not the only one with issues https://askubuntu.com/questions/492269/microsoft-sculpt-ergonomic-desktop-virtual-joystick-how-to-remove-it

im not sure how to make a work around for it.  given that it is an issue affecting just your keyboard, do you mind if i open a new issue for it so we can move this forward?

**Comment by bitcraft** [comment]
Opened #567

**Comment by kerizane** [comment]
If you go to the monster menu and then press "Shift" the following error happens:

Traceback (most recent call last):
  File "tuxemon.py", line 50, in <module>
    tuxemon.core.main.main(load_slot=args.slot)
  File "/space/git/Tuxemon/tuxemon/core/main.py", line 106, in main
    control.main()
  File "/space/git/Tuxemon/tuxemon/core/control.py", line 250, in main
    update(clock_tick)
  File "/space/git/Tuxemon/tuxemon/core/control.py", line 290, in update
    events = list(self.process_events(events))
  File "/space/git/Tuxemon/tuxemon/core/control.py", line 198, in process_events
    game_event = self._send_event(game_event)
  File "/space/git/Tuxemon/tuxemon/core/control.py", line 217, in _send_event
    game_event = state.process_event(game_event)
  File "/space/git/Tuxemon/tuxemon/core/components/menu/menu.py", line 433, in process_event
    disabled = all(not i.enabled for i in self.menu_items)
AttributeError: 'MonsterMenuState' object has no attribute 'menu_items'

**Comment by bitcraft** [comment]
Fixed the shift key issue

**Comment by kerizane** [comment]
Looks good.
Would like to have another look this evening.

**Comment by kerizane** [comment]
Shift key behaves surprisingly when naming your character. This is pre-existing though.

**Comment by bitcraft** [comment]
opened an issue for the shift key.  i've also implemented a blacklist for joystick devices.  i'd like to draw the line here for related fixes if possible.

**Pull request merged**